src/window.c

```diff
@@ -6314,6 +6334,104 @@ set_fraction(win_T *wp)
+ * TODO: Ensure this also works with wrapped lines.
+ * Requires topline to be able to be set to a bufferline
+ * with some offset(row-wise scrolling/smoothscroll).
+ */
+    static void
+win_fix_scroll(win_T *next_curwin, int close)
+{
+    int      ls_offset = (p_ls == 1 && win_count() == 2) ? 1 : 0;
+    win_T    *wp;
+    linenr_T lnum;
+
+    FOR_ALL_WINDOWS(wp)
+    {
+	// Skip when window height has not changed or when
+	// buffer has less lines than the window height.
+	if (wp->w_height != wp->w_prev_height
+		&& wp->w_height < wp->w_buffer->b_ml.ml_line_count)
+	{
+	    lnum = wp->w_cursor.lnum;
+	    // Determine cursor position required to keep the same screenlines.
+	    if (wp->w_winrow != wp->w_prev_winrow)
+	    { // For bottom windows, keep the same botline except for 'ls' == 1.
+		if (!close && ((wp->w_winrow + VISIBLE_HEIGHT(wp)
+					+ wp->w_status_height) == cmdline_row))
+		    wp->w_cursor.lnum = wp->w_botline - 1 - ls_offset;
+		else // Otherwise determine based on winrow and height.
+		    wp->w_cursor.lnum = MIN(wp->w_botline - 1
+					    + (wp->w_winrow - wp->w_prev_winrow)
+					    + (wp->w_height - wp->w_prev_height),
+					    wp->w_buffer->b_ml.ml_line_count);
+		// Bring the new cursor position to the bottom of the screen.
+		wp->w_fraction = FRACTION_MULT;
+		scroll_to_fraction(wp, wp->w_prev_height);
+		wp->w_cursor.lnum = lnum;
+	    }
+	    invalidate_botline_win(wp);
+	    validate_botline_win(wp);
+	    // Ensure cursor position is valid for the current window to be
+	    // or if currently not in normal/cmdline mode.
+	    if (wp == curwin && !(get_real_state() & (MODE_NORMAL|MODE_CMDLINE)))
+		win_fix_cursor(wp, TRUE);
+	    else if (wp == next_curwin)
+		win_fix_cursor(wp, FALSE);
+	}
+	wp->w_prev_height = wp->w_height;
+	wp->w_prev_winrow = wp->w_winrow;
+    }
+}
+
+/*
+ * Make sure cursor position is valid for 'nosplitscroll'.
+ * If it is not, save cursor position to jumplist and move it.
+ * If we are not in normal mode, scroll to make valid instead.
+ */
+    static void
+win_fix_cursor(win_T *wp, int normal)
+{
+    long     so = wp->w_p_so < 0 ? p_so : wp->w_p_so;
+    win_T    *cw = curwin;
+    buf_T    *cb = curbuf;
+    linenr_T nlnum = 0;
+
+    if (wp->w_buffer->b_ml.ml_line_count < wp->w_height)
+	return;
+
+    so = MIN(wp->w_height / 2, so);
+    // Check if cursor position is above topline or below botline.
+    if (wp->w_cursor.lnum < (wp->w_topline + so) && wp->w_topline != 1)
+	nlnum = MIN(wp->w_topline + so, wp->w_buffer->b_ml.ml_line_count);
+    else if (wp->w_cursor.lnum > (wp->w_botline - so - 1)
+	    && (wp->w_botline - wp->w_buffer->b_ml.ml_line_count) != 1)
+	nlnum = MAX(wp->w_botline - so - 1, 1);
+    // If cursor was invalid scroll or change cursor.
+    if (nlnum)
+    {   // Ensure cursor stays visible if we are not in normal mode.
+	if (normal)
+	{
+	    wp->w_fraction = FRACTION_MULT;
+	    scroll_to_fraction(wp, wp->w_prev_height);
+	}
+	else
+	{
+	    curwin = wp;
```

Comment:
@brammool I think this workaround is fine. Keeps the problem/solution isolated and it works properly. I guess if there ever is a need to add a setmark_win() in the future it could be used here too but for now I would do it like this.